File: frontend/lib/data/services/calendar_service.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'api_service.dart';
import '../models/calendar_event.dart';
import '../models/marketplace.dart';

class CalendarService {
  final ApiService _api;

  CalendarService(this._api);

  Future<List<CalendarEvent>> getEvents({DateTime? startDate, DateTime? endDate}) async {
    final response = await _api.get('/calendar/events', queryParameters: {
      if (startDate != null) 'startDate': startDate.toIso8601String(),
      if (endDate != null) 'endDate': endDate.toIso8601String(),
    });
    return (response.data as List).map((e) => CalendarEvent.fromJson(e)).toList();
  }

  Future<CalendarEvent> createEvent(CalendarEvent event) async {
    final response = await _api.post('/calendar/events', data: event.toJson());
    return CalendarEvent.fromJson(response.data);
  }

  Future<CalendarEvent> updateEvent(String eventId, Map<String, dynamic> data) async {
    final response = await _api.put('/calendar/events/$eventId', data: data);
    return CalendarEvent.fromJson(response.data);
  }

  Future<void> deleteEvent(String eventId) async {
    await _api.delete('/calendar/events/$eventId');
  }

  Future<Booking> createBooking(String serviceId, DateTime date) async {
    final response = await _api.post('/calendar/bookings', data: {
      'serviceId': serviceId,
      'date': date.toIso8601String(),
    });
    return Booking.fromJson(response.data);
  }

  Future<List<Booking>> getMyBookings() async {
    final response = await _api.get('/calendar/bookings/my');
    return (response.data as List).map((e) => Booking.fromJson(e)).toList();
  }

  Future<List<Booking>> getBusinessBookings() async {
    final response = await _api.get('/calendar/bookings/business');
    return (response.data as List).map((e) => Booking.fromJson(e)).toList();
  }
}

final calendarServiceProvider = Provider<CalendarService>((ref) {
  final apiService = ApiService();
  return CalendarService(apiService);
});

File: backend/src/services/calendarService.ts
import prisma from '../config/database';

export class CalendarService {
  // Получение событий
  static async getEvents(userId: string, startDate?: Date, endDate?: Date) {
    const where: any = { userId };
    if (startDate && endDate) {
      where.date = {
        gte: startDate,
        lte: endDate,
      };
    }

    return prisma.calendarEvent.findMany({
      where,
      orderBy: { date: 'asc' },
    });
  }

  // Получение события по ID
  static async getEventById(eventId: string, userId: string) {
    const event = await prisma.calendarEvent.findUnique({
      where: { id: eventId },
    });

    if (!event || event.userId !== userId) {
      throw new Error('Событие не найдено');
    }

    return event;
  }

  // Создание события
  static async createEvent(userId: string, data: any) {
    return prisma.calendarEvent.create({
      data: { ...data, userId },
    });
  }

  // Обновление события
  static async updateEvent(eventId: string, userId: string, data: any) {
    const event = await prisma.calendarEvent.findUnique({
      where: { id: eventId },
    });

    if (!event || event.userId !== userId) {
      throw new Error('Событие не найдено или нет прав');
    }

    return prisma.calendarEvent.update({
      where: { id: eventId },
      data,
    });
  }

  // Удаление события
  static async deleteEvent(eventId: string, userId: string) {
    const event = await prisma.calendarEvent.findUnique({
      where: { id: eventId },
    });

    if (!event || event.userId !== userId) {
      throw new Error('Событие не найдено или нет прав');
    }

    await prisma.calendarEvent.delete({ where: { id: eventId } });
    return { success: true };
  }

  // Записи на услуги
  static async createBooking(serviceId: string, clientId: string, date: Date) {
    return prisma.booking.create({
      data: {
        serviceId,
        clientId,
        date,
        status: 'pending',
      },
      include: {
        service: true,
        client: {
          select: {
            id: true,
            firstName: true,
            lastName: true,
            phone: true,
          },
        },
      },
    });
  }

  static async getClientBookings(clientId: string) {
    return prisma.booking.findMany({
      where: { clientId },
      include: {
        service: {
          include: {
            user: {
              select: {
                id: true,
                businessName: true,
                avatar: true,
              },
            },
          },
        },
      },
      orderBy: { date: 'asc' },
    });
  }

  static async getBusinessBookings(userId: string) {
    return prisma.booking.findMany({
      where: {
        service: { userId },
      },
      include: {
        service: true,
        client: {
          select: {
            id: true,
            firstName: true,
            lastName: true,
            phone: true,
            avatar: true,
          },
        },
      },
      orderBy: { date: 'asc' },
    });
  }
}
